SSA/Core/Framework.lean

```diff
@@ -220,282 +590,717 @@ args, since there now are equation lemmas for it.
+  ((Com.ret (Op:=Op) (eff := eff) v).changeVars varsMap).outContext = Γ' := by simp
+
+@[simp] lemma Com.denote_changeVars
+    (varsMap : Γ.Hom Γ') (c : Com Op Γ eff ty) :
+    (c.changeVars varsMap).denote =
+    c.denote ∘ (fun V => V.comap varsMap) := by
+  funext Γ'v
+  simp only [Function.comp_apply]
+  induction c using Com.rec' generalizing Γ'v Γ' with
+  | ret x => simp [Com.denote, Com.changeVars, Valuation.comap, *]
+  | lete _ _ ih =>
+    unfold denote
+    simp only [changeVars_lete]
+    split <;> simp_all
+
+@[simp] lemma Com.denote_changeVars' (varsMap : Γ.Hom Γ') (c : Com Op Γ eff ty) :
+    (c.changeVars varsMap).denote = fun V => c.denote (V.comap varsMap) := by
+  simp; rfl
+
+@[simp] lemma Com.outContext_changeVars_hom {map : Γ.Hom Δ} (map_inv : Δ.Hom Γ) :
+    {c : Com Op Γ eff ty} → Ctxt.Hom (outContext (changeVars c map)) (outContext c)
+  | .ret _        => cast (by simp) map_inv
+  | .lete _ body  => cast (by simp) <|
+      Com.outContext_changeVars_hom (map := map.snocMap) map_inv.snocMap (c := body)
+
+-- @[simp] lemma Com.denoteLets_changeVars (map : Γ.Hom Δ) (c : Com Op Γ .pure ty) :
+--     (c.changeVars map).denoteLets = fun (V : Valuation Δ) =>
+--       (c.denoteLets (V.comap map)).comap (fun t' v' => v'.castCtxt (by simp)) := by
+--   sorry
+
+@[simp] lemma Com.denoteLets_returnVar_pure (c : Com Op Γ .pure ty) (Γv : Valuation Γ) :
+    c.denoteLets Γv c.returnVar = c.denote Γv := by
+  induction c using Com.recPure <;> simp_all [denoteLets, denote]
+
+-- @[simp] lemma Com.comap_denoteLets_pure [LawfulMonad m]
+--     (map : Γ.Hom Δ) (c : Com Op Γ .pure ty) (Δv : Valuation Δ) :
+--     Valuation.comap ((c.changeVars map).denoteLets Δv)
+--       (c.changeVars map).outContextHom = Δv := by
+--   funext t' v'
+--   simp only [Valuation.comap]
+--   induction c using Com.recPure generalizing Δ Δv
+--   case ret => rfl
+--   case lete ih =>
+--     stop
+--     rw [changeVars_lete, outContextHom_lete];
+--     simp
+
+@[simp] lemma Expr.changeVars_changeVars (e : Expr Op Γ eff ty) (f : Γ.Hom Δ) (g : Δ.Hom Ξ) :
+    (e.changeVars f).changeVars g = e.changeVars (f.comp g) := by
+  rcases e with ⟨op, ty_eq, eff_le, args, regArgs⟩
+  simp [changeVars, HVector.map_map]
+  rfl
+
+/-!
+## FlatCom
+An alternative representation of a program as a `Lets` with a return `Var`
+-/
 
-def Com.changeVars
-    (varsMap : Γ.Hom Γ') :
-    Com Op Γ ty → Com Op Γ' ty
-  | .ret e => .ret (varsMap e)
-  | .lete e body => .lete (e.changeVars varsMap)
-      (body.changeVars (fun t v => varsMap.snocMap v))
+/-- `FlatCom Γ eff Δ ty` represents a program as a sequence `Lets Γ eff Δ` and a `Var Δ ty`.
+This is isomorphic to `Com Γ eff ty`, where `Δ` is `com.outContext` -/
+structure FlatCom (Op : _) {Ty : _} [OpSignature Op Ty m]  (Γ_in : Ctxt Ty) (eff : EffectKind)
+    (Γ_out_new : Ctxt Ty) (ty : Ty) where
+--TODO: `Γ_out_new` should be named just `Γ_out`
```

Comment:
Why is this a TODO and not a change? What's the issue with doing it?

Reply:
done.